.github/workflows/elixir.yml

```diff
@@ -12,60 +12,159 @@ concurrency:
       - uses: asdf-vm/actions/install@v1
         if: steps.asdf-cache.outputs.cache-hit != 'true'
-      # if we did hit the cache, set up the environment
-      - name: Setup ASDF environment
-        run: |
-          echo "ASDF_DIR=$HOME/.asdf" >> $GITHUB_ENV
-          echo "ASDF_DATA_DIR=$HOME/.asdf" >> $GITHUB_ENV
-        if: steps.asdf-cache.outputs.cache-hit == 'true'
-      - name: Reshim ASDF
-        run: |
-          echo "$ASDF_DIR/bin" >> $GITHUB_PATH
-          echo "$ASDF_DIR/shims" >> $GITHUB_PATH
-          $ASDF_DIR/bin/asdf reshim
+      - run: |
+          mix local.rebar --force
+          mix local.hex --force
+        if: steps.asdf-cache.outputs.cache-hit != 'true'
+
+  formatting:
+    name: Formatting
+    runs-on: ubuntu-latest
+    needs: asdf
+
+    steps:
+      - uses: actions/checkout@v3
+      - name: ASDF cache
+        uses: actions/cache@v3
+        with:
+          path: ~/.asdf
+          key: ${{ runner.os }}-asdf-v3-${{ hashFiles('.tool-versions') }}
+      - uses: mbta/actions/reshim-asdf@v1
+      - name: Check formatting
+        run: mix format --check-formatted
+
+  deps:
+    name: Get Dependencies
+    runs-on: ubuntu-latest
+    needs: asdf
+
+    steps:
+      - uses: actions/checkout@v3
       - name: Restore dependencies cache
         id: deps-cache
-        uses: actions/cache@v2
+        uses: actions/cache@v3
         with:
           path: deps
-          key: ${{ runner.os }}-mix-${{ hashFiles('**/mix.lock') }}
+          key: ${{ runner.os }}-mix-v2-${{ hashFiles('**/mix.lock') }}
           restore-keys: ${{ runner.os }}-mix-
+      - name: ASDF cache
+        uses: actions/cache@v3
+        with:
+          path: ~/.asdf
+          key: ${{ runner.os }}-asdf-v3-${{ hashFiles('.tool-versions') }}
+        if: steps.deps-cache.outputs.cache-hit != 'true'
+      - uses: mbta/actions/reshim-asdf@v1
+        if: steps.deps-cache.outputs.cache-hit != 'true'
       - name: Install dependencies
-        run: |
-          mix local.rebar --force
-          mix local.hex --force
-          mix deps.get
+        run: mix deps.get
+        if: steps.deps-cache.outputs.cache-hit != 'true'
+
+  dev:
+    name: Dev Tasks
+    runs-on: ubuntu-latest
+    needs: deps
+
+    steps:
+      - uses: actions/checkout@v3
+      - name: ASDF cache
+        uses: actions/cache@v3
+        with:
+          path: ~/.asdf
+          key: ${{ runner.os }}-asdf-v3-${{ hashFiles('.tool-versions') }}
+      - uses: mbta/actions/reshim-asdf@v1
+      - name: Restore dependencies cache
+        id: deps-cache
+        uses: actions/cache@v3
+        with:
+          path: deps
+          key: ${{ runner.os }}-mix-v2-${{ hashFiles('**/mix.lock') }}
+          restore-keys: ${{ runner.os }}-mix-v2-
+      - name: Restore dev build cache
+        uses: actions/cache@v3
+        with:
+          path: _build/dev
+          key: ${{ runner.os }}-mix-dev-${{ hashFiles('**/mix.lock') }}
```

Comment:
Seems like we'll end up restoring old results in this directory. Is the idea that reusing those is still probably faster than a full rebuild?

Reply:
Yeah; like with the build directory on your local machine, even if you update some dependencies, most of the _build/dev directory is still used for the unmodified deps.

Reply:
But doesn't `_build/dev` also contain built artifacts of the application when you run `mix run` or `mix compile`?

Reply:
It does, yes. But as with your local machine, Elixir will recompile files that have been modified since the cache was updated.